### DISCUSSION TITLE
map_async for generated data?

### DISCUSSION THREAD

**Discussion by erlendp** [gfx-rs/wgpu#2413] [Q&A]
Should I favour `BufferSlice::map_async` when I need to update a buffer regularly with data generated on the fly, as presumably this will allow me to avoid allocating a temporary CPU buffer, or would I be better to just use `Queue::write_buffer`?

**Comment by cwfitzgerald** [comment]
map_async carries with it quite a bit of complexity, and write_buffer is designed with being a fast-path in mind (that also happens to be very easy to use)

**Reply by erlendp** [reply]
So it's safe to say that writing my data to an in memory data structure first, and then calling `write_buffer` will generally outperform `map_async`, followed by writing directly to a device buffer?

P.S. If the answer is "probably, but you should profile", then that's good enough for me!

**Reply by kvark** [reply]
A well-written `map_async` is able to be faster than a well-written `write_buffer` (on native only). It's especially visible in cases like video uploads - you have a concrete known size each frame, and re-using buffers of this size is straightforward. For random-sized updates it's more of a tie, there are small pros/cons. Generally, don't bother with `map_async` until you are doing an optimization and figuring out the hot spots.
